pig/src/org/partiql/pig/domain/parser/TypeDomainParser.kt

```diff
@@ -15,275 +15,350 @@
 
 package org.partiql.pig.domain.parser
 
-import com.amazon.ionelement.api.*
 import com.amazon.ion.IonReader
 import com.amazon.ion.system.IonReaderBuilder
-import org.partiql.pig.domain.model.*
+import com.amazon.ionelement.api.AnyElement
+import com.amazon.ionelement.api.IonElement
+import com.amazon.ionelement.api.IonElementException
+import com.amazon.ionelement.api.IonElementLoaderOptions
+import com.amazon.ionelement.api.IonLocation
+import com.amazon.ionelement.api.MetaContainer
+import com.amazon.ionelement.api.SexpElement
+import com.amazon.ionelement.api.SymbolElement
+import com.amazon.ionelement.api.createIonElementLoader
+import com.amazon.ionelement.api.emptyMetaContainer
+import com.amazon.ionelement.api.head
+import com.amazon.ionelement.api.location
+import com.amazon.ionelement.api.metaContainerOf
+import com.amazon.ionelement.api.tag
+import com.amazon.ionelement.api.tail
+import org.partiql.pig.domain.model.Arity
+import org.partiql.pig.domain.model.DataType
+import org.partiql.pig.domain.model.NamedElement
+import org.partiql.pig.domain.model.PermutedDomain
+import org.partiql.pig.domain.model.PermutedSum
+import org.partiql.pig.domain.model.Statement
+import org.partiql.pig.domain.model.Transform
+import org.partiql.pig.domain.model.TupleType
+import org.partiql.pig.domain.model.TypeDomain
+import org.partiql.pig.domain.model.TypeRef
+import org.partiql.pig.domain.model.TypeUniverse
+import org.partiql.pig.errors.ErrorContext
+import java.io.File
+import java.io.FileNotFoundException
+import java.io.InputStream
+import java.util.Stack
+
+
+fun parseTypeUniverseFile(path: String): TypeUniverse {
+    val parser = Parser(FILE_SYSTEM_SOURCE)
+    return parser.parseTypeUniverse(path)
+}
+
+internal class Parser(
+    val inputSource: InputSource
+) {
+    /** This contains every file the parser has "seen" and is used detect and prevent import cycles. */
+    private val parseHistory = HashSet<String>()
+    private val qualifedSourceStack = Stack<String>().apply { push(".") }
+
+    /** Parses a type universe in the specified [IonReader]. */
+    fun parseTypeUniverse(source: String): TypeUniverse {
+        val elementLoader = createIonElementLoader(IonElementLoaderOptions(includeLocationMeta = true))
+        val qualifiedSource = inputSource.getCanonicalName(source)
+        return inputSource.openStream(qualifiedSource).use { inputStream: InputStream ->
```

Comment:
I agree with Almann on having a resolver to take care of the imports. Also, the `HashSet` doesn't necessarily avoid cyclic dependencies TBH. It just makes sure that the file imported earlier is not imported again. Consider this scenario for example, 
`universe_a.ion` contains a `type a` depending on `universe_b.ion`
```
(import universe_b) 
...
(sum expr
     ...
     (type_a value::type_b)
     ...
```
and `universe_b.ion` contains a `type b` depending on `universe_a.ion`
```
(import universe_a) 
...
(sum expr
     ...
     (type_b value::type_a)
     ...
```
 NOTE: My PIG grammar may be wrong but you get the point. 
 
 In the above example, `type_a` or `type_b` are not really defined as they depend on each other and hence there's a cyclic dependency. We really should be throwing an error in this case. 
 On that note, I don't think the current implementation handles this. 
 Also, how are you handling the naming clashes if there are multiple sources imported having the same type name?

Reply:
If I understand the issue you're trying to describe here...

> Also, the HashSet doesn't necessarily avoid cyclic dependencies TBH. 

You're correct that this PR doesn't address cyclic dependencies between types and type domains, however that's because this PR actually has nothing to do with that.  The `HashSet` *does* in fact prevent cyclic _imports/includes_ from being a problem, and that's all it does.  

> Consider this scenario for example,

This should demonstrate why that's not a valid scenario with the PIG of today:

```
(define domain_a
    (domain 
        (product type_a symbol)
    )
)
(define domain_b
    (domain 
        (product type_b type_a) 
        //              ^ PIG issues an error here since type_a does not exist in domain_b!  
    )
)
```

Until #65 is completed, there is *no* way for `domain_b` to reference `type_a`.  Given this, it's not possible to have such a cyclic dependency.  

Should we want to add a static check that makes cyclic dependencies such as described in your scenario an error, we will do that while #65 is being implemented.